### PR TITLE
Add heating

### LINKED ISSUE #106: Add heating
Options will include
- Start temperature
- Final temperature
- Temperature step size
- Number of steps, or more likely time, at each temperature

This will (at least initially) be implemented via additional options in md, similarly to [lavello](https://gitlab.com/drFaustroll/lavello/-/blob/main/md.py).

### PR DESCRIPTION
**Pull request by ElliottKasoar**
Resolves #106

To do:
- [x] More/better tests including using the data reader

### CODE REVIEW

File: janus_core/cli.py [hunk 1]
```diff
@@ -667,6 +667,20 @@ def md(
...
+        Optional[float],
+        typer.Option(help="Temperature to start heating, in K.  [default: None]"),
+    ] = None,
+    temp_end: Annotated[
+        Optional[float],
+        typer.Option(help="Maximum temperature for heating, in K.  [default: None]"),
+    ] = None,
+    temp_step: Annotated[
+        float, typer.Option(help="Size of temperature steps when heating, in K.")
+    ] = 10,
```
Comment (alinelena) [inline janus_core/cli.py:680]: def
```suggestion
    ] = None,
```

File: janus_core/cli.py [hunk 2]
```diff
@@ -667,6 +667,20 @@ def md(
...
+    temp_time: Annotated[
+        float, typer.Option(help="Time between heating steps, in fs.")
+    ] = 10,
```
Comment (alinelena) [inline janus_core/cli.py:683]:
```suggestion
    ] = None,
```

File: janus_core/md.py [hunk 3]
```diff
@@ -153,6 +161,10 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
         traj_append: bool = False,
         traj_start: int = 0,
         traj_every: int = 100,
+        temp_start: Optional[float] = None,
+        temp_end: Optional[float] = None,
+        temp_step: float = 10,
```
Comment (alinelena) [inline janus_core/md.py:166]:
```suggestion
        temp_step: Optional[float] = None,
```

File: janus_core/md.py [hunk 4]
```diff
@@ -153,6 +161,10 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
+        temp_time: float = 10,
```
Comment (alinelena) [inline janus_core/md.py:167]:
```suggestion
        temp_time: Optional[float] = None,
```

File: janus_core/md.py [hunk 5]
```diff
@@ -216,6 +228,16 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
             Step to start saving trajectory. Default is 0.
         traj_every : int
             Frequency of steps to save trajectory. Default is 100.
+        temp_start : Optional[float]
+            Temperature to start heating, in K. Default is None, which disables
+            heating.
+        temp_end : Optional[float]
+            Maximum temperature for heating, in K. Default is None, which disables
+            heating.
+        temp_step : float
```
Comment (alinelena) [inline janus_core/md.py:237]:
```suggestion
        temp_step : Optional[float]
```

File: janus_core/md.py [hunk 6]
```diff
@@ -216,6 +228,16 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
+            Size of temperature steps when heating, in K. Default is 10.
+        temp_time : float
```
Comment (alinelena) [inline janus_core/md.py:239]:
```suggestion
        temp_time : Optional[float]
```

File: tests/test_md.py [hunk 7]
```diff
@@ -525,3 +528,94 @@ def test_atoms_struct(tmp_path):
...
+    single_point = SinglePoint(
+        struct_path=DATA_PATH / "NaCl.cif",
+        architecture="mace",
+        calc_kwargs={"model": MODEL_PATH},
+    )
+
+    with pytest.raises(ValueError):
+        NVT(
+            struct=single_point.struct,
+            temp=300.0,
```
Comment (alinelena) [inline tests/test_md.py:581]:
```suggestion
```

File: janus_core/cli.py [hunk 8]
```diff
@@ -749,6 +763,16 @@ def md(
...
+    temp_start : Optional[float]
+        Temperature to start heating, in K. Default is None, which disables
+        heating.
+    temp_end : Optional[float]
+        Maximum temperature for heating, in K. Default is None, which disables
+        heating.
+    temp_step : float
+        Size of temperature steps when heating, in K. Default is 10.
+    temp_time : float
+        Time between heating steps, in fs. Default is 10.
```
Comment (ElliottKasoar) [inline janus_core/cli.py:775]:
```suggestion
    temp_step : Optional[float]
        Size of temperature steps when heating, in K. Default is None, which disables
        heating.
    temp_time : Optional[float]
        Time between heating steps, in fs. Default is None, which disables
        heating.
```

File: janus_core/md.py [hunk 9]
```diff
@@ -85,6 +85,14 @@ class MolecularDynamics:  # pylint: disable=too-many-instance-attributes
...
     traj_every : int
         Frequency of steps to save trajectory. Default is 100.
+    temp_start : Optional[float]
+        Temperature to start heating, in K. Default is None, which disables heating.
+    temp_end : Optional[float]
+        Maximum temperature for heating, in K. Default is None, which disables heating.
+    temp_step : float
+        Size of temperature steps when heating, in K. Default is 10.
+    temp_time : float
+        Time between heating steps, in fs. Default is 10.
```
Comment (ElliottKasoar) [inline janus_core/md.py:95]:
```suggestion
    temp_step : Optional[float]
        Size of temperature steps when heating, in K. Default is None, which disables heating.
    temp_time : Optional[float]
        Time between heating steps, in fs. Default is None, which disables heating.
```

File: janus_core/md.py [hunk 10]
```diff
@@ -216,6 +228,16 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
         traj_every : int
             Frequency of steps to save trajectory. Default is 100.
+        temp_start : Optional[float]
+            Temperature to start heating, in K. Default is None, which disables
+            heating.
+        temp_end : Optional[float]
+            Maximum temperature for heating, in K. Default is None, which disables
+            heating.
+        temp_step : float
+            Size of temperature steps when heating, in K. Default is 10.
```
Comment (ElliottKasoar) [inline janus_core/md.py:238]:
```suggestion
            Size of temperature steps when heating, in K. Default is None, which disables heating.
```

File: janus_core/md.py [hunk 11]
```diff
@@ -216,6 +228,16 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
+        temp_time : float
+            Time between heating steps, in fs. Default is 10.
```
Comment (ElliottKasoar) [inline janus_core/md.py:240]:
```suggestion
            Time between heating steps, in fs. Default is None, which disables heating.
```

File: janus_core/md.py [hunk 12]
```diff
@@ -289,25 +321,29 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
-    def rotate_restart_files(self) -> None:
+    def _rotate_restart_files(self) -> None:
         """Rotate restart files."""
         if len(self.restart_files) > self.restarts_to_keep:
             path = Path(self.restart_files.pop(0))
             path.unlink(missing_ok=True)
 
-    def reset_velocities(self) -> None:
+    def _set_velocity_distribution(self) -> None:
         """Reset velocities and (optionally) rotation of system."""
```
Comment (alinelena) [inline janus_core/md.py:331]:
```suggestion
        """Set velocities to current target temperature and (optionally) rotation of system."""
```

**Review by alinelena — DISMISSED** [review]

File: janus_core/md.py [hunk 13]
```diff
@@ -289,25 +321,29 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals,too-many-sta
...
             path = Path(self.restart_files.pop(0))
             path.unlink(missing_ok=True)
 
-    def reset_velocities(self) -> None:
-        """Reset velocities and (optionally) rotation of system."""
-        if self.dyn.nsteps < self.equil_steps:
-            MaxwellBoltzmannDistribution(self.struct, temperature_K=self.temp)
-            Stationary(self.struct)
+    def _set_velocity_distribution(self) -> None:
+        """Set velocities to current target temperature and (optionally) rotation of system."""
```
Comment (ElliottKasoar) [inline janus_core/md.py:331]:
```suggestion
        """
        Set velocities to current target temperature.
        
        Sets Maxwell-Boltzmann velocity distribution, as well as removing
        centre-of-mass momentum, and (optionally) total angular momentum.
        """
```

**Review by alinelena — APPROVED** [review]

**Pull request merged**